### PR TITLE
Data structure

### PR DESCRIPTION
**Pull request by zetraxus**

### CODE REVIEW

File: src/data_structure/board.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package data_structure;
+
+import java.util.Vector;
+
+public class board {
+    private byte width;
+    private byte height;
+
+    private field2D[][] gameMap;
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
    private field2D[][] gameBoard;
```

File: src/data_structure/board.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    private Vector<fieldInfo> columnsInfo;
+    private Vector<fieldInfo> rowsInfo;
+
+    public board(byte width, byte height, field2D[] startMap) {
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
    public board(byte width, byte height, field2D[] startBoard) {
```

File: src/data_structure/board.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        gameMap = new field2D[width][height];
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
        gameBoard = new field2D[width][height];
```

File: src/data_structure/board.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        columnsInfo = new Vector<>();
+        rowsInfo = new Vector<>();
+
+        for (int i = 0; i < height; ++i)
+            for (int j = 0; j < width; ++j) {
+                gameMap[j][i] = startMap[i * width + j];
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                gameBoard[j][i] = startBoard[i * width + j];
```

File: src/data_structure/board.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+                if (gameMap[j][i].getType() == field2D.Type.INFOCOLUMN || gameMap[j][i].getType() == field2D.Type.INFOCOLUMNANDROW)
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                if (gameBoard[j][i].getType() == field2D.Type.INFOCOLUMN || gameBoard[j][i].getType() == field2D.Type.INFOCOLUMNANDROW)
```

File: src/data_structure/board.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                    columnsInfo.add(gameMap[j][i].getColumn());
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                    columnsInfo.add(gameBoard[j][i].getColumn());
```

File: src/data_structure/board.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                if (gameMap[j][i].getType() == field2D.Type.INFOROW || gameMap[j][i].getType() == field2D.Type.INFOCOLUMNANDROW)
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                if (gameBoard[j][i].getType() == field2D.Type.INFOROW || gameBoard[j][i].getType() == field2D.Type.INFOCOLUMNANDROW)
```

File: src/data_structure/board.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                    columnsInfo.add(gameMap[j][i].getRow());
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                    columnsInfo.add(gameBoard[j][i].getRow());
```

File: src/data_structure/board.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            }
+
+        for (int i = 0; i < columnsInfo.size(); ++i) {
+            byte x = columnsInfo.elementAt(i).getX();
+            byte y = (byte) (columnsInfo.elementAt(i).getY() + 1);
+
+            for (int j = 0; j < columnsInfo.elementAt(i).getFieldCount(); ++j, ++y) {
+                gameMap[x][y].getWritable().setColumnFieldInfo(columnsInfo.elementAt(i));
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                gameBoard[x][y].getWritable().setColumnFieldInfo(columnsInfo.elementAt(i));
```

File: src/data_structure/board.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+                columnsInfo.elementAt(i).addWritableFields(gameMap[x][y].getWritable());
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                columnsInfo.elementAt(i).addWritableFields(gameBoard[x][y].getWritable());
```

File: src/data_structure/board.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            }
+        }
+
+        for (int i = 0; i < rowsInfo.size(); ++i) {
+            byte x = (byte) (rowsInfo.elementAt(i).getX() + 1);
+            byte y = rowsInfo.elementAt(i).getY();
+
+            for (int j = 0; j < rowsInfo.elementAt(i).getFieldCount(); ++j, ++x) {
+                gameMap[x][y].getWritable().setRowFieldInfo(rowsInfo.elementAt(i));
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                gameBoard[x][y].getWritable().setRowFieldInfo(rowsInfo.elementAt(i));
```

File: src/data_structure/board.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                rowsInfo.elementAt(i).addWritableFields(gameMap[x][y].getWritable());
```
Comment (PROZ-283766) [inline src/data_structure/board.java]:
```suggestion
                rowsInfo.elementAt(i).addWritableFields(gameBoard[x][y].getWritable());
```

File: src/data_structure/board.java (hunk 2 above)
Comment (kam193) [inline src/data_structure/board.java]: Why `startMap` is only 1D array? I think 2D array should be better and simpler to use.

File: src/data_structure/board.java (hunk 1 above)
Comment (kam193) [reply src/data_structure/board.java]: I agree, rename `Map` to `Board`

File: src/data_structure/field2D.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
+package data_structure;
+
+public class field2D { // we need that class to gather information about rows and column in one fieldWritable
+
+    private fieldInfo column;
+    private fieldInfo row;
+    private fieldWritable writable;
+    private Type type;
+
+    public fieldInfo getColumn() {
```
Comment (kam193) [inline src/data_structure/field2D.java:33]: I think, we need a constructor, setter or default values to can fill this fields.

File: src/data_structure/fieldInfo.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
+package data_structure;
+
+import java.util.Vector;
+
+public class fieldInfo {
+
+    private byte x, y; // location field on gameMap
+    private byte sum;
+    private byte fieldCount; // count of fields which should sum to variable sum in this class
+    private Direction direction;
```
Comment (kam193) [inline src/data_structure/fieldInfo.java]: I think we don't need direction now.

File: src/data_structure/fieldInfo.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
+package data_structure;
+
+import java.util.Vector;
+
+public class fieldInfo {
+
+    private byte x, y; // location field on gameMap
+    private byte sum;
+    private byte fieldCount; // count of fields which should sum to variable sum in this class
```
Comment (kam193) [inline src/data_structure/fieldInfo.java:9]: Again: we need a constructor or setters to can fill this fields.

File: src/data_structure/fieldInfo.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+    public byte getFieldCount() {
+        return fieldCount;
+    }
+
+    public Direction getDirection() {
+        return direction;
+    }
+
+    public void addWritableFields(fieldWritable fieldWritable) {
```
Comment (kam193) [inline src/data_structure/fieldInfo.java:36]: Please add also a getter to can get connected fields

File: src/data_structure/fieldWritable.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
+package data_structure;
+
+public class fieldWritable {
+
+    private boolean[] possibilities;
+    private byte possibilitiesCount;
+    private State state;
```
Comment (kam193) [inline src/data_structure/fieldWritable.java:9]: Please add setters/getters/another method to manipulate for fields.

**Review by kam193 — CHANGES_REQUESTED** [review]

**Review by kam193 — APPROVED** [review]

**Pull request merged**